jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/FileMappedHttpContentFactory.java

```diff
@@ -0,0 +1,124 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2022 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.http;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Objects;
+
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.thread.AutoLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FileMappedHttpContentFactory implements HttpContent.Factory
+{
+    private static final Logger LOG = LoggerFactory.getLogger(FileMappedHttpContentFactory.class);
+    private static final int DEFAULT_MIN_FILE_SIZE = 16 * 1024;
+
+    private final HttpContent.Factory _factory;
+    private final int _minFileSize;
+
+    /**
+     * Construct a {@link FileMappedHttpContentFactory} which can use file mapped buffers.
+     * Uses a default value of {@value DEFAULT_MIN_FILE_SIZE} for the minimum size of an
+     * {@link HttpContent} before trying to use a file mapped buffer.
+     *
+     * @param factory the wrapped {@link HttpContent.Factory} to use.
+     */
+    public FileMappedHttpContentFactory(HttpContent.Factory factory)
+    {
+        this(factory, DEFAULT_MIN_FILE_SIZE);
+    }
+
+    /**
+     * Construct a {@link FileMappedHttpContentFactory} which can use file mapped buffers.
+     *
+     * @param factory the wrapped {@link HttpContent.Factory} to use.
+     * @param minFileSize the minimum size of an {@link HttpContent} before trying to use a file mapped buffer.
+     */
+    public FileMappedHttpContentFactory(HttpContent.Factory factory, int minFileSize)
+    {
+        _factory = Objects.requireNonNull(factory);
+        _minFileSize = minFileSize;
+    }
+
+    @Override
+    public HttpContent getContent(String path) throws IOException
+    {
+        HttpContent content = _factory.getContent(path);
+        if (content != null && content.getContentLengthValue() > _minFileSize)
+            return new FileMappedContent(content);
```

Comment:
So we don't map the buffer here because we are not sure if something will call `getByteBuffer`?
What exactly is the setup that could cause that to happen?

Reply:
We don't map it here because it's a heavy operation that blocks and takes time, which we don't want to spend in the supposedly fast and light `getHttpContent()`.

Reply:
This feels so wrong.  The fast path here is for a file that has already been mapped long ago, so we should just be able to use it directly without a lock and probably without even a volatile lookup.

Firstly are we really sure that mapping a file is a heavy operation?  I would have thought it would be cheaper than reading the file as it just makes an entry into the memory map of the process.   But let's assume that it is expensive....

Then the cache/factory should be able to deal with this in one of several ways:
 1. just block and map it.  Let's see what the cost of this slow path is.
 2. return the original HC until such time as the mapped buffer is loaded, at which time the entry in the cache is replaced with the fast mapped HC
 3. return a HC that waits (preferably threadlessly... but that is another issue) until the cache has been updated with the fast mapped HC.

Reply:
Note that when an `HttpContent` is cached it is wrapped by `CachedHttpContent` which will extract out the buffer using `getByteBuffer()`. It is then stored as a final field in `CachedHttpContent` and then the usage of this cached version never hits the volatile just the final field.

If you do not use the `CachingHttpContentFactory`, then using `FileMappedHttpContent.getBuffer()` directly will be hitting this volatile and potentially locking to generate the mapped ByteBuffer.

Reply:
Good point, except now we have just transferred the delay/work from the `getByteBuffer` call on the FMHC to the `getHttpContent` on the CCF.    I.e any work done in getting the buffer in a HC that might be cached will end up being work done in the `getHttpContent` on the cache, so we have the same global lock and/or stampede problems there.

Note that I'm a bit dubious if we should ever map a file that is not being cached, as it is a big investment to make for a once off file.

I'm starting to think that file mapping should actually just be an aspect of the cache, that when it decides to load a resource it can do so either as a mapped file or as a byte buffer.  The cache entry can initially be one that either blocks; or calls the non cached HC;  or eventually queues requests until the real HC is ready and replaces it in the cache.

Reply:
File mapping should not be a cache issue. 
If I want to map a file because an external process also maps it and changes it, I would like to return a mapped buffer even if there is no cache in the front.

Reply:
> we have just transferred the delay/work from the getByteBuffer call on the FMHC to the getHttpContent on the CCF. I.e any work done in getting the buffer in a HC that might be cached will end up being work done in the getHttpContent on the cache, so we have the same global lock and/or stampede problems there.

Because we do the file mapping in `getByteBuffer()` we don't have a stampede of threads all mapping the file. 

Each thread gets a `FileMappedHttpContent` from the factory, they all try to add them to the cache, but the creation of the `CachingHttpContent` (which calls `getByteBuffer()`) is done atomically inside the `_cache.computeIfAbsent` so only one thread will ever do the buffer mapping of the file, the rest of the `FileMappedHttpContent` instances will be discarded because they did not win the race to add it into the cache and they will not also map the file because `getByteBuffer()` was never called.

Reply:
Maybe, but that would be yet another case where the implementation of `getByteBuffer` has to somehow telepathically know the intentions of the caller  - does it really want a mapped buffer or not?  is it calling it to cache it or not?  It's just another indication that we don't have the API exactly right yet... together with the fact that we've just had several cycles debating if it is best to do the work in the getContent or the getBuffer, only to realise that if it is cached then all the work is done in the getContent anyway!    We just can't tell why `getByteBuffer` is being called, so we can't optimise it or apply any reasonable policy that holds generally.  

There is a lot of goodness in this PR and I think the things I don't like are due to the API not being correct yet.   I think we need to get the design right (in another PR) and then focus back in on micro optimising the implementations.  So let's get this merged and then consider the API.

Reply:
> Each thread gets a `FileMappedHttpContent` from the factory, they all try to add them to the cache, but the creation of the `CachingHttpContent` (which calls `getByteBuffer()`) is done atomically inside the `_cache.computeIfAbsent` so only one thread will ever do the buffer mapping of the file, the rest of the `FileMappedHttpContent` instances will be discarded because they did not win the race to add it into the cache and they will not also map the file because `getByteBuffer()` was never called.

And this is yet another example of a fragile design.   The cache is only not stampeding because we "known" that all the expensive stuff is done in `getByteBuffer`.

I'd much rather a cache design where the element initially put into the cache could make an explicit choice when other requests arrived for the same resource to block them waiting; queue them asynchronously; or dispatch them to the uncached content.     We would have explicit behavior encoded rather than emergent behavior from disciplined implementations of cheap constructors and expensive getBuffers.

Reply:
> Maybe, but that would be yet another case where the implementation of getByteBuffer has to somehow telepathically know the intentions of the caller

That is not true. The implementation does not care at all. It just returns a mapped buffer.
Whether it is the same one mapped once, or many mapped multiple times, the caller does not know and does not care.
You can argue that it is not efficient to map multiple times, but it has nothing to do with the API design.

> We just can't tell why getByteBuffer is being called, so we can't optimise it or apply any reasonable policy that holds generally.

I still don't understand what you are talking about here.
Not a single implementation cares about why or when `getByteBuffer()` is being called.
It is called, it returns a `ByteBuffer` if it can, done.

I still think we are looking for problems that don't exist?

> I'd much rather a cache design where the element initially put into the cache could make an explicit choice when other requests arrived for the same resource to block them waiting; queue them asynchronously; or dispatch them to the uncached content.

We have said repeatedly that this is a better job for a 3rd party cache implementation, not us.

Reply:
> > Maybe, but that would be yet another case where the implementation of getByteBuffer has to somehow telepathically know the intentions of the caller
> 
> That is not true. The implementation does not care at all. It just returns a mapped buffer. 

Exactly - a "mapped buffer" because it some how has to know that "an external process also maps it and changes it".

We should just have a HC that serves content as best it can.  If the caller wants a more efficient mechanism, then it is their responsibility to buffer that content or perhaps create a MappedFIleHC.

 
> We have said repeatedly that this is a better job for a 3rd party cache implementation, not us.

Except that we don't have an API that is powerful enough to use a 3rd party cache.  We couldn't use anything that needs to look at cache-control headers for example, as they are not available in getByteBuffer?


I don't understand why you are fighting against this proposal so hard, it can do everything that the getByteBuffer style can and I'm not proposing that initially we change much from what we are currently doing.  But by having a fully powerful API, it allows us to do more powerful things in the future.   It also means that we leverage our existing produce API rather than invent yet another API?

Reply:
Closing this discussion to move it elsewhere and not distract from this PR.